synapse/handlers/sliding_sync.py

```diff
@@ -260,49 +462,87 @@ async def current_sync_for_user(
                         sync_config.user, sync_room_map, list_config.filters, to_token
                     )
 
+                # Sort the list
                 sorted_room_info = await self.sort_rooms(
                     filtered_sync_room_map, to_token
                 )
 
+                # Find which rooms are partially stated and may need to be filtered out
+                # depending on the `required_state` requested (see below).
+                partial_state_room_map = await self.store.is_partial_state_room_batched(
+                    filtered_sync_room_map.keys()
+                )
+
+                # Since creating the `RoomSyncConfig` takes some work, let's just do it
+                # once and make a copy whenever we need it.
+                room_sync_config = RoomSyncConfig.from_room_config(list_config)
+
                 ops: List[SlidingSyncResult.SlidingWindowList.Operation] = []
                 if list_config.ranges:
                     for range in list_config.ranges:
-                        sliced_room_ids = [
-                            room_id
-                            # Both sides of range are inclusive
-                            for room_id, _ in sorted_room_info[range[0] : range[1] + 1]
-                        ]
+                        room_ids_in_list = []
+
+                        # We're going to loop through the sorted list of rooms starting
+                        # at the range start index and keep adding rooms until we fill
+                        # up the range or run out of rooms.
+                        #
+                        # Both sides of range are inclusive
+                        current_range_index = range[0]
+                        range_end_index = range[1]
+                        while (
+                            current_range_index <= range_end_index
+                            and current_range_index <= len(sorted_room_info) - 1
+                        ):
+                            room_id, _ = sorted_room_info[current_range_index]
```

Comment:
Would this not be easier done as:

```suggestion
                        num_rooms = range[1] - range[0] + 1
                        for room_id, _ in sorted_room_info[range[0]:]:
                            if len(room_ids_in_list) >= num_rooms:
                                break
```

Reply:
Looks easier to follow :+1: